### PR TITLE
Vis3r code

### PR DESCRIPTION
**Pull request by TheArchKnight**
This PR is created for adding support for the NxtVR device to the official OpenHMD repo. Code had been tested by me and it works really well (used the godot ohmd example). We would like to have our driver for the DIY HMD we are creating. You can visit our GitHub for see all of our progress.

Thanks for your attention,
Anorak
2019

### CODE REVIEW

**Comment by TheOnlyJoey** [comment]
Please remove the configure.ac and makefile.am from the build, since this is removed after 0.3 release from OpenHMD. just Meson and CMake are supported as build systems.

**Comment by AdmiraalHugo** [comment]
Alright, I have removed the files you listed. Both checks still fail to succeed successfully. Can you please have a look at it, @TheOnlyJoey ?

File: src/drv_nxtvr/nxtvr.c [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+/*  Nxtvr Driver */
+
+//Using as code base: https://github.com/der-b/OpenHMD/tree/master/src/drv_sparkfun_9dof
+
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <hidapi/hidapi.h>
+#include <openhmdi.h>
```
Comment (TheOnlyJoey) [inline src/drv_nxtvr/nxtvr.c]: If you check the compilation logs in travis and appvoyeur you can see that openhmdi.h is not found.
https://travis-ci.org/OpenHMD/OpenHMD/jobs/588911091

If you reference the other drivers, you see that this library is only included in the header files of the drivers with relative path. (the library is also already included in the nxtvr.h)
Please remove this line.

File: src/drv_nxtvr/nxtvr.c [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        {
+            LOGE("NXTVR: Invalid data size %d for reportID 42.", size);
+            return;
+        };
+
+        quatf quater;
+        //copy remaining 16 bytes into float union
+        memcpy((void *)&quater, (void *)&buffer[1], 16);
+
+        priv->base.rotation = quater;
```
Comment (TheOnlyJoey) [inline src/drv_nxtvr/nxtvr.c]: This seems to use pre-calculated quaternions, where OpenHMD prefers to use raw accel and gyro data from the IMU to be able to use our own Sensor Fusion and allow for more and easier optimization.

Also using memcpy to get data from your buffer into the driver quaternion is bad practice.
A preferred method is to use the read functions that we implemented in most drivers.
https://github.com/OpenHMD/OpenHMD/blob/master/src/drv_nolo/packet.c#L97

File: src/drv_nxtvr/nxtvr.c [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+{
+    nxt_priv *priv = ohmd_alloc(driver->ctx, sizeof(nxt_priv));
+
+    if (!priv)
+        return NULL;
+
+    priv->handle = hid_open_path(desc->path);
+    if (!priv->handle)
+    {
+        LOGD("Clould not open directory '" DEVICE_NAME "'.");
```
Comment (TheOnlyJoey) [inline src/drv_nxtvr/nxtvr.c]: Typo, i think you it should be "Could?"

File: src/drv_nxtvr/nxtvr.c [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        goto err;
+    }
+
+    if (hid_set_nonblocking(priv->handle, 1) == -1)
+    {
+        LOGD("Clould not set nonblocking on '" DEVICE_NAME "'.");
```
Comment (TheOnlyJoey) [inline src/drv_nxtvr/nxtvr.c]: Typo again

File: src/drv_nxtvr/nxtvr.c [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        goto err_handle;
+    }
+
+    // Set default device properties
+    ohmd_set_default_device_properties(&priv->base.properties);
+
+    // Set device properties (imitates the rift values)
```
Comment (TheOnlyJoey) [inline src/drv_nxtvr/nxtvr.c]: Due to this data being a rift copy, is this driver even ready?
If this is still in development i would recommend maintaining this in your own repository until the headset is finalized, before adding this into Master.

File: src/drv_nxtvr/nxtvr.c (hunk 1 above)
Comment (AdmiraalHugo) [reply src/drv_nxtvr/nxtvr.c]: fixed.

File: src/drv_nxtvr/nxtvr.c (hunk 4 above)
Comment (AdmiraalHugo) [reply src/drv_nxtvr/nxtvr.c]: fixed the typo

File: src/drv_nxtvr/nxtvr.c (hunk 3 above)
Comment (AdmiraalHugo) [reply src/drv_nxtvr/nxtvr.c]: fixed the typo

**Comment by TheOnlyJoey** [comment]
Closing the PR for now. This looks like a incomplete driver for a incomplete project. Would be happy to have this in once the HMD is completed and functional.

For now I would just work on this on your own fork. The OpenHMD shared object can be used in any application using OpenHMD due to the compatible ABI so its not a drawback for compatibility.

**Pull request closed without merging**